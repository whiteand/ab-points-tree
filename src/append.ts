import { tree as getTree } from './tree';
import { ITree } from './types';

function _append<Point extends number[]>(
  tree: ITree<Point> | null,
  point: Point,
): ITree<Point> {
  if (!tree) {
    const ranges = point.map(value => [value, value] as [number, number]);
    return getTree(point, ranges, null, null);
  }

  const headX = tree.point[0];
  const pointX = point[0];
  const ranges = Array(point.length);

  if (headX === pointX) {
    ranges[0] = tree.ranges[0];

    for (let i = 1; i < point.length; i++) {
      const leftRange = tree.left ? tree.left.ranges[i] : null;
      const rightRange = tree.right ? tree.right.ranges[i] : null;
      const y = point[i];
      let minY = y;
      let maxY = y;

      if (leftRange) {
        if (leftRange[0] < minY) {
          minY = leftRange[0];
        }
        if (leftRange[1] > maxY) {
          maxY = leftRange[1];
        }
      }

      if (rightRange) {
        if (rightRange[0] < minY) {
          minY = rightRange[0];
        }
        if (rightRange[1] > maxY) {
          maxY = rightRange[1];
        }
      }

      ranges[i] = [minY, maxY];
    }

    return getTree(point, ranges, tree.left, tree.right);
  }
  if (headX > pointX) {
    const newLeft = append(tree.left, point);

    for (let i = 0; i < point.length; i++) {
      let minY = tree.point[i];
      let maxY = tree.point[i];
      const leftRange = newLeft.ranges[i];
      const rightRange = tree.right ? tree.right.ranges[i] : null;
      if (leftRange[0] < minY) {
        minY = leftRange[0];
      }
      if (leftRange[1] > maxY) {
        maxY = leftRange[1];
      }
      if (rightRange) {
        if (rightRange[0] < minY) {
          minY = rightRange[0];
        }
        if (rightRange[1] > maxY) {
          maxY = rightRange[1];
        }
      }
      ranges[i] = [minY, maxY];
    }

    return getTree(tree.point, ranges, newLeft, tree.right);
  }
  const newRight = append(tree.right, point);

  for (let i = 0; i < point.length; i++) {
    let minY = tree.point[i];
    let maxY = tree.point[i];
    const leftRange = tree.left ? tree.left.ranges[i] : null;
    const rightRange = newRight.ranges[i];
    if (leftRange) {
      if (leftRange[0] < minY) {
        minY = leftRange[0];
      }
      if (leftRange[1] > maxY) {
        maxY = leftRange[1];
      }
    }
    if (rightRange[0] < minY) {
      minY = rightRange[0];
    }
    if (rightRange[1] > maxY) {
      maxY = rightRange[1];
    }
    ranges[i] = [minY, maxY];
  }

  return getTree(tree.point, ranges, tree.left, newRight);
}

export function append<Point extends number[]>(
  tree: ITree<Point> | null,
  ...points: Point[]
): ITree<Point> {
  let res = _append(tree, points[0]);
  for (let i = 1; i < points.length; i++) {
    const point = points[i];
    res = _append(res, point);
  }
  return res;
}
