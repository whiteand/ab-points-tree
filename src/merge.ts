import { append } from './append';
import { Queue } from './queue';
import { ITree } from './types';

export function merge<Point extends number[]>(
  t1: ITree<Point> | null,
  t2: ITree<Point> | null,
) {
  if (!t1) {
    return t2;
  }
  if (!t2) {
    return t1;
  }

  const queue = new Queue<ITree<Point>>();

  queue.push(t2);

  let newTree = t1;

  while (queue.length > 0) {
    const t = queue.pop();

    newTree = append(newTree, t.point);

    if (t.left) {
      queue.push(t.left);
    }
    if (t.right) {
      queue.push(t.right);
    }
  }

  return newTree;
}
