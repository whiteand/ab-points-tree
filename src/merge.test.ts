import { append } from './append';
import { merge } from './merge';

describe('merge function', () => {
  test('merge(null, null)', () => {
    expect(merge(null, null)).toBe(null);
  });
  test('merge(tree, null)', () => {
    const t1 = append(null, [1, 2]);
    expect(merge(t1, null)).toBe(t1);
  });
  test('merge(null, tree)', () => {
    const t2 = append(null, [2, 4]);
    expect(merge(null, t2)).toBe(t2);
  });
  test('merge of two trees', () => {
    const t1 = append(null, [4, 6], [2, 4], [6, 8]);
    const t2 = append(null, [5, 7], [3, 5], [7, 9]);
    const merged = merge(t1, t2);
    expect(merged).toMatchSnapshot();
  });
});
