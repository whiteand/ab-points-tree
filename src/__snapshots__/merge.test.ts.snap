// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge function merge of two trees 1`] = `
Object {
  "left": Object {
    "left": null,
    "point": Array [
      2,
      4,
    ],
    "ranges": Array [
      Array [
        2,
        3,
      ],
      Array [
        4,
        5,
      ],
    ],
    "right": Object {
      "left": null,
      "point": Array [
        3,
        5,
      ],
      "ranges": Array [
        Array [
          3,
          3,
        ],
        Array [
          5,
          5,
        ],
      ],
      "right": null,
    },
  },
  "point": Array [
    4,
    6,
  ],
  "ranges": Array [
    Array [
      2,
      7,
    ],
    Array [
      4,
      9,
    ],
  ],
  "right": Object {
    "left": Object {
      "left": null,
      "point": Array [
        5,
        7,
      ],
      "ranges": Array [
        Array [
          5,
          5,
        ],
        Array [
          7,
          7,
        ],
      ],
      "right": null,
    },
    "point": Array [
      6,
      8,
    ],
    "ranges": Array [
      Array [
        5,
        7,
      ],
      Array [
        7,
        9,
      ],
    ],
    "right": Object {
      "left": null,
      "point": Array [
        7,
        9,
      ],
      "ranges": Array [
        Array [
          7,
          7,
        ],
        Array [
          9,
          9,
        ],
      ],
      "right": null,
    },
  },
}
`;
